### PR TITLE
Release automation

### PR DESCRIPTION
**Pull request by malaskowski**
## Description
First step in making [release automation](https://github.com/Knotx/knotx-aggregator/issues/19) more pleasant. Most of work done by @dragarth-7

## Motivation and Context
Migrate building to Gradle in order to unify build tools across all Knot.x official repositories.

## Upgrade notes (if appropriate)
Insteag `mvn package` run `./gradlew builld` in order to build a new image.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2019 Knot.x Project
```
Comment (tomaszmichalak) [inline build.gradle.kts:2]:
```suggestion
 * Copyright (C) 2020 Knot.x Project
```

Please fix also other files :)

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+plugins {
+    id("com.bmuschko.docker-remote-api")
+    id("org.nosphere.apache.rat")
+}
+
+repositories {
+    jcenter()
+    mavenLocal()
+    maven { url = uri("https://plugins.gradle.org/m2/") }
+    maven { url = uri("https://oss.sonatype.org/content/groups/staging/") }
```
Comment (tomaszmichalak) [inline build.gradle.kts]: Do we need the staging and snapshots repositories?

File: build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    register("build") {
+        group = "build"
+        dependsOn("downloadAndUnzipDistribution", "prepareDocker")
+    }
+
+    register("clean") {
+        group = "build"
+        dependsOn("removeBaseImage", "removeBaseAlpineImage")
+        doFirst {
```
Comment (tomaszmichalak) [inline build.gradle.kts]: Maybe `doLast` ?

File: src/main/docker/base/Dockerfile [hunk 4]
```diff
@@ -1,3 +1,17 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
 # A base Dockerfile for Knot.x
 
 FROM openjdk:8-jre
```
Comment (tomaszmichalak) [inline src/main/docker/base/Dockerfile:17]: LABEL maintainer="Knot.x Project"

File: gradle/docker.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+tasks.register<DockerBuildImage> ("buildBaseAlpineImage") {
+    group = "docker"
+    inputDir.set(file("$buildDir/out"))
+    dockerFile.set(file("$buildDir/out/baseAlpine/Dockerfile"))
+    images.add(dockerBaseAlpineImageId)
+    dependsOn("removeBaseAlpineImage", "copyDockerfileBaseAlpine")
+}
+
+tasks.register<DockerSaveImage>("saveBaseImage") {
```
Comment (tomaszmichalak) [inline gradle/docker.gradle.kts:83]: I am wondering if we can:
- define `base` and `alpine` as variations of tasks, what if we add some `example` in the future
- then we can define variations' list and register tasks groups per variation

**Review by tomaszmichalak — CHANGES_REQUESTED** [review]

File: build.gradle.kts (hunk 1 above)
Comment (malaskowski) [reply build.gradle.kts:2]: The year in the license headers is not specified by the AL standards.
AFAIK the good practice is to leave the header as the year of the project/module inception. Which is `2019` for knotx-docker.

File: build.gradle.kts (hunk 2 above)
Comment (malaskowski) [reply build.gradle.kts]: If we want to be able to build local SNAPSHOT of a docker image - yes.
Maybe staging is not required.

File: src/main/docker/base/Dockerfile (hunk 4 above)
Comment (malaskowski) [reply src/main/docker/base/Dockerfile:17]: It's there, just GitHub is not presenting the full file:
https://github.com/Knotx/knotx-docker/blob/feature/release-automation/src/main/docker/base/Dockerfile#L18

File: gradle/docker.gradle.kts (hunk 5 above)
Comment (malaskowski) [reply gradle/docker.gradle.kts:83]: I was thinking the same, `remove` task has logic unified into the Kotlin method.
However, tasks like `saveImage` or `buildImage` have every configuration value different, so even when we abstract their logic, still there will be the same numbers of params to configure.
I came to the conclusion that this form is easier to understand.
Do you see any particular example to extract logic to the more general task/method?

File: build.gradle.kts (hunk 3 above)
Comment (malaskowski) [reply build.gradle.kts]: I've applied [`base`](https://docs.gradle.org/current/userguide/base_plugin.html) plugin instead.

**Review by tomaszmichalak — APPROVED** [review]

**Pull request merged**